### ISSUE TITLE
World.collect() drops the reset frame and never produces boundary NaN actions

### ISSUE THREAD

**Issue by Matteodambr** [galilai-group/stable-worldmodel#293]
Hi all,
Found this issue while working with some custom gym datasets, and i took the liberty of proposing a PR as well aimed at fixing this behavior.

## Summary of bug

`World.collect()` doesn't record the observation from `env.reset()`. Frame 0 of every recorded episode is actually the observation after the *first* `.step()` call, not the reset observation. Because of that, the whole episode ends up shifted by one step, and the code that's supposed to write a NaN action at the episode boundary ends up writing nothing.

## Location

`stable_worldmodel/world/world.py` (0.1.1), from the `World.collect()` / `_run_iter` rewrite in #197.

`collect()` buffers frames through an `on_step` callback:

```python
def on_step(world):
    for col, data in world.infos.items():
        ...
        buffers[i][col].append(val)
...
for env_idx, _ in self._run_iter(
    episodes=episodes, seed=seed, options=options,
    mode='auto', on_step=on_step,
):
    ep = {k: list(v) for k, v in buffers[env_idx].items()}
    buffers[env_idx].clear()
    if 'action' in ep:
        ep['action'].append(ep['action'].pop(0))
    ...
```

but in `_run_iter`, `on_step` only gets called after `.step()`:

```python
if seed is not None or options is not None:
    self.reset(seed=seed, options=options)          # no on_step call

...
for t in range(...):
    ...
    self.envs.step(actions, mask=mask)
    if on_step:
        on_step(self)                                # only fires here
    ...
    if mode == 'auto':
        ...
        self.envs.reset(seed=seeds, options=options, mask=done)   # no on_step call
        ...
```

Neither the initial reset nor the per-env auto-reset ever triggers `on_step`, so the buffer for each episode starts one step late.

That also breaks the `ep['action'].append(ep['action'].pop(0))` line right above — it's left over from the old `record_dataset()`, which buffered the reset frame first and had a comment explaining the shift (`# left-shift actions to align with observations i.e. (o_t, a_t)`). Since the reset frame isn't buffered anymore, that shift now pops the real first action and sticks it at the end instead of a NaN.

The wrapper still does the right thing on its side, `MegaWrapper` sets `info['action']` to NaN on reset and to the real action on step (`stable_worldmodel/wrapper/default.py`). `collect()` just never sees the reset call.

## Relevance

`scripts/train/pldm.py` expects that NaN-at-boundary convention:

```python
batch['action'] = torch.nan_to_num(batch['action'], 0.0)  # "occurs at sequence boundaries"
...
output['idm_emb'] = torch.cat([emb[:, 1:], emb[:, :-1]], dim=-1)
output['act_label'] = batch['action'][:, :-1].detach()
```

Since the collected data is shifted by one step, `action[t]` next to `obs[t]` is actually the action that produced `obs[t]`, not the action from `obs[t]` to `obs[t+1]`. So the inverse-dynamics labels used in training are off by one, for every window, not just at episode boundaries. Nothing crashes and there's no NaN in the loss (the `nan_to_num` call hides it), so this would just quietly make the model worse rather than throwing an error. `tests/data/test_integration_real.py` is the only test that runs `collect()` end to end and it only checks episode count, keys, and dtypes, so nothing catches this.

## Reproducibility

```python
from stable_worldmodel import World
from stable_worldmodel.policy import RandomPolicy
import numpy as np

w = World(env_name='swm/TwoRoom-v1', num_envs=1, image_shape=(32, 32),
          max_episode_steps=6, render_target=True)
w.set_policy(RandomPolicy(seed=0))

buffers = [dict(state=[]) for _ in range(1)]
def on_step(world):
    buffers[0]['state'].append(np.array(world.infos['state'][0]).copy())

w.reset(seed=0)
reset_state = np.array(w.infos['state'][0]).copy()
print('reset state (expected as buffered_states[0]):', reset_state)

for i, ep in w._run_iter(episodes=1, seed=0, on_step=on_step):
    print('buffered states:', len(buffers[0]['state']))
    for s in buffers[0]['state']:
        print(' ', s)
w.close()
```

Output:
```
reset state (expected as buffered_states[0]): [[159.05367   87.527466]]
buffered states: 6
  [[160.42328   85.225334]]   # this is the state after step 1, not the reset state
  [[155.83301  80.39061]]
  [[158.96571   84.518166]]
  [[160.03207  86.81313]]
  [[160.46832  91.16386]]
  [[163.62686  86.19124]]
```

The reset state never shows up in the buffer, regardless of episode length. Same story if you write to an actual file: collecting one episode with `World.collect()` (TwoRoom-v1, `max_episode_steps=15`, HDF5 writer) gives you a 15-step episode with no NaN anywhere in the `action` array.

## Fix

Have `on_step` fire once after both reset points in `_run_iter`, the initial reset and the per-env auto-reset in the `mode == 'auto'` branch, so the buffer gets the reset frame before any step happens. For the auto-reset case it should probably only fire for the envs that were actually reset (the `done` mask), otherwise you'd double-count envs that are still mid-episode. Once that's in place, the existing pop/append shift in `collect()` should work as originally intended.

**Comment by quentinll** [comment]
Hey @Matteodambr, thank you for reporting this and opening a PR! We will have a look asap

**Comment by Matteodambr** [comment]
Thanks @quentinll ! Just to give a bit more context, my concern is that this issue generates an actual misalignment in terms of timesteps between the action and the observations seen by the models, for every episode.

The gymnasium convention for stepping is: `state_t, action_t --step(action_t)--> state_{t+1}`, which is also what the actual rollout loop in `World._run_iter` implements: the policy computes `a_t = policy.get_action(self.infos)` using the current `self.infos` (i.e. `s_t`) before `self.infos` gets overwritten by `.step()`. So `a_t` is chosen from `s_t` and produces `s_{t+1}` correctly.

Prior to the fix, `on_step` was never called at reset, so buffer row 0 was actually the info dict form the first `.step()`, not the reset. The wrapper records `info['action']` after stepping (`EverythingToInfoWrapper.step` in `default.py`), so raw buffered row `i` held `(pixels_{i+1}, action_{i+1})`, which is just missing the reset frame entirely. `collect()`'s rotation (`ep['action'].append(ep['action'].pop(0))`) then shifts the `action` column forward by one, under the assumption that row 0 holds the reset's NaN action. Since that NaN was never buffered, the rotation instead shifted a real action into that slot, misaligning every row by one:

| disk row `i` | `pixels` | `action` | should have been |
|---|---|---|---|
| 0 | `pixels_1` | `action_2` | `action_1` |
| 1 | `pixels_2` | `action_3` | `action_2` |
| ... | ... | ... | ... |
| n-2 | `pixels_{n-1}` | `action_n` | `action_{n-1}` |
| n-1 | `pixels_n` | `action_1` | NaN |

So every `pixels_t` was paired with `action_{t+1}`, the action from one step later than it should be. The terminal row got a valid-looking action (`action_1`) instead of the NaN. With `history_size`/`num_preds=1` in the JEPA training loop, this meant each prediction target's causing action was sitting one row ahead of where the model actually looked for it.

Post-fix, `on_step` now also fires after both reset points (initial + per-env auto-reset), so the true reset frame is buffered as row 0 with its NaN action. The same rotation now correctly shifts that NaN to the tail instead of a real action:

| disk row `i` | `pixels` | `action` |
|---|---|---|
| 0 | `pixels_0` (reset) | `action_0` |
| 1 | `pixels_1` | `action_1` |
| ... | ... | ... |
| n-1 | `pixels_{n-1}` | `action_{n-1}` |
| n | `pixels_n` | NaN |

Every row now holds `(pixels_t, action_t)` with matching indices, matching the gym convention.

For example, see `dinowm_forward` in `scripts/train/prejepa.py:96-109`:

```python
def dinowm_forward(self, batch, stage, cfg):
    ...
    batch = self.model.encode(
        batch,
        target='emb',
        is_video=cfg.backbone.get('is_video_encoder', False),
    )

    embedding = batch['emb'][:, : cfg.wm.history_size, ...]
    pred_embedding = self.model.predict(embedding)
    target_embedding = batch['emb'][:, cfg.wm.num_preds :, ...].detach()
```

`self.model.encode` bundles `pixels_emb[t]` and `action_emb[t]` together into `batch['emb'][t]` at the same row index (see `PreJEPA.encode` in `stable_worldmodel/wm/prejepa/prejepa.py:28-67`, and `Embedder.forward` in `stable_worldmodel/wm/prejepa/module.py:97-102`, which just tiles the raw `action[t]` column alongside `pixels[t]` with no relative shift). So context row `t` is `(pixels_t, action_t)`, and with `num_preds=1` the target for that row is `target_embedding[t] = batch['emb'][t+1]`, i.e. `pixels_{t+1}`.

Pre-fix, `action[t]` on disk was actually `action_{t+1}` (per the misalignment table above), the action that *produces* `pixels_{t+1}`, i.e. exactly the target the loss at that row is scoring against.
